cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -223,8 +223,13 @@ data ProjectBuildContext = ProjectBuildContext {
 runProjectPreBuildPhase
     :: Verbosity
     -> ProjectBaseContext
-    -> (ElaboratedInstallPlan -> IO ElaboratedInstallPlan)
-    -> IO ProjectBuildContext
+    -- ^ A function that selects a subset of the plan and returns it,
+    -- It can also return some arbitrary data, which can be useful in case
+    -- the caller needs it.
+    -> (ElaboratedInstallPlan -> IO (ElaboratedInstallPlan, a))
```

Comment:
> Note that in that case we still have to modify all call sites to return `Nothing`.

Hmm, I wonder if we should stick the whole `TargetMap` inside `ProjectBuildContext`. That would be potentially more useful to non-`new-run` clients (e.g. `new-bench`/`new-test`, where I think we want the ability to run several benchmarks/tests at once), and we could set it to something meaningful other than `Nothing` in those cases.

We'd need to do the check that there's only a single target twice in `new-run` code, though (makes sense to add a helper for this).

> You mean the return type? 

Yep, the definition of `ProjectBuildContext`.